src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala

```diff
@@ -171,18 +171,18 @@ class GoogleExtensions(
         directoryDAO.getSynchronizedDate(id).map(dateOption => id -> dateOption)
       }
       // only sync groups that have already been synchronized
-      messagesForIdsWithSyncDates = idsAndSyncDates.collect {
-        case (gn: WorkbenchGroupName, Some(_)) => gn.toJson.compactPrint
-        case (rpn: FullyQualifiedPolicyId, Some(_)) => rpn.toJson.compactPrint
-      }
-      _ <- IO.fromFuture(IO(googlePubSubDAO.publishMessages(googleServicesConfig.groupSyncTopic, messagesForIdsWithSyncDates)))
-      ancestorGroups <- groupIdentities.toList.traverse { id =>
+      managedGroupNames = idsAndSyncDates.collect { case (groupName: WorkbenchGroupName, Some(_)) => groupName }
+      accessPolicyIds   = idsAndSyncDates.collect { case (accessPolicyId: FullyQualifiedPolicyId, Some(_)) => accessPolicyId }
+
+      _ <- IO.fromFuture(IO(googlePubSubDAO.publishMessages(googleServicesConfig.groupSyncTopic, managedGroupNames.map(_.toJson.compactPrint) ++ accessPolicyIds.map(_.toJson.compactPrint))))
+
+      ancestorManagedGroups <- managedGroupNames traverse { id =>
         directoryDAO.listAncestorGroups(id)
       }
-      managedGroupIds = (ancestorGroups.flatten ++ groupIdentities).filterNot(visitedGroups.contains).collect {
+      managedGroupIds = (ancestorManagedGroups.flatten ++ groupIdentities).filterNot(visitedGroups.contains).collect {
         case FullyQualifiedPolicyId(FullyQualifiedResourceId(ManagedGroupService.managedGroupTypeName, id), ManagedGroupService.adminPolicyName | ManagedGroupService.memberPolicyName) => id
       }
-      _ <- managedGroupIds.traverse(id => onManagedGroupUpdate(id, visitedGroups ++ groupIdentities ++ ancestorGroups.flatten))
+      _ <- managedGroupIds.traverse(id => onManagedGroupUpdate(id, visitedGroups ++ groupIdentities ++ ancestorManagedGroups.flatten))
```

Comment:
I think the changes in this function do what we want but I would like to clean it up a little more. The main thing is that the recursion is unnecessary. But removing that means a bigger restructuring.

I feel like there are 2 steps, 1) figure out what messages to send and 2) send them. The first step has 2 branches: if it is a policy just create a single message for the policy but if it is a managed group, find any related policies via auth domains of the group or an ancestor, create a message for each policy and a message for the managed group. I think that code will end up being more straight forward.

Reply:
Ah - I didn't realize we were getting rid of the recursion - I'll change it!